Cache/Config.lean

```diff
@@ -0,0 +1,84 @@
+import Lean.Data.RBMap
+import Lean.Data.Json
+
+open System (FilePath mkFilePath)
+open Lean (Json RBMap)
+
+namespace Cache
+
+abbrev PkgDirs := Lean.RBMap String FilePath compare
+
+structure Config where
+  url : String
+  head : FilePath
+  rootDir : FilePath
+  tokenEnvVar : String
+  pkgDirs: PkgDirs
+
+namespace Config
+
+@[inline] def getBuildDir (cfg : Config) (k : String) : Option FilePath :=
+  cfg.pkgDirs.find? k
+
+def FILE : FilePath :=
+  ⟨"cache-config.json"⟩
+
+def stdTemplate : String :=
+"{
+  \"url\":\"https://lakecache.blob.core.windows.net/std4\",
+  \"head\":\"Std.lean\",
+  \"rootDir\":\"lake-packages/std\",
+  \"tokenEnvVar\":\"STD_CACHE_SAS\",
+  \"pkgDirs\":
+  {
+    \"Std\":\"lake-packages/std\"
+  }
```

Comment:
`rootDir`, `pkgDirs`, and `head` should be autodetected.  We're probably going to add additional dependencies to mathlib at some point in the future.  And that shouldn't break everyone's configuration.

Reply:
Not having dependencies listed in `pkgDirs` doesn't break `cache`. It simply ignores them and doesn't cache their build files.

Reply:
> Not having dependencies listed in pkgDirs doesn't break cache. 

That's effectively the same.  If they are not listed, then they won't be cached and large parts of mathlib will need to be recompiled every time (because they depend on the not-cached files).

Reply:
> How to autodetect those?

Good question.

The `pkgDirs` is easy.  We could just read `lake-manifest.json`.

For the rest, I think we need to parse the lakefile.  Lake has a `Lake.loadWorkspace` function for that.

Reply:
I think we will need to add `Lake` as a dependency in order to be able to do that. Can we get away with just figuring out `pkgDirs` and having the other fields to be explicitly stated in the config file?

Reply:
Oof, we can `import Lake` but can't link against it. :sigh:

Alternatively/additionally could we maybe just have the config file say "take all settings from `lake-packages/mathlib/cache-config.json`"?

Reply:
Not all settings can be taken from that file. In particular, `rootDir` and the value for the `Mathlib` key in the `pkgDirs` map will be different. In `mathlib4` they point to `.` whereas in files that depend on `mathlib4` they point to `lake-packages/<MATHLIB>`, where "MATHLIB" is some custom dir name they can tell Lake to clone `mathlib4` to.

Check this out: https://leanprover.zulipchat.com/#narrow/stream/287929-mathlib4/topic/how.20to.20use.20the.20mathlib.20cache.3F/near/329078991

Reply:
The `rootDir` is easy: it's just the directory that the `cache-config.json` file is in.

> is some custom dir name they can tell Lake to clone mathlib4 to.

I don't think you can change that directory name, it's always going to be `mathlib`.